### PR TITLE
Add on demand assignment provider pallet

### PR DESCRIPTION
**Pull request by antonva**
All runtimes except Rococo start to use the parachain assignment pallet as their assignment provider while Rococo uses a multiplexing assignment provider that provides both on demand and parachain assignments.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2707996

File: runtime/parachains/src/assigner_on_demand.rs [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+		// Get the position of the next `ParaId`. Select either a `ParaId` that has an affinity
+		// to the same `CoreIndex` as the scheduler asks for or a `ParaId` with no affinity at all.
+		let pos = queue.iter().position(|entry| match ParaIdAffinity::<T>::get(&entry.claim.0) {
+			Some(affinity) => return affinity.core_idx == core_idx,
+			None => return true,
+		});
+
+		if let Some(i) = pos {
+			// Smell, should either check pos/i or the result of remove, not both.
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand.rs:395]: You can do an: pos.and_then(|p| queue.remove(p)

File: runtime/parachains/src/assigner_on_demand.rs [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+				// Write changes to storage.
+				OnDemandQueue::<T>::set(queue);
+				return Some(Assignment::ParathreadA(entry))
+			};
+		}
+		return None
+	}
+
+	/// Pushes an assignment back to the queue on session boundaries.
+	/// Drops the assignment if retries exceeds `config.parathread_retries`.
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand.rs:407]: Retries should not concern us here.

File: runtime/parachains/src/assigner_on_demand.rs [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+				Pallet::<T>::increase_affinity(entry.claim.0, core_idx);
+				// Write changes to storage.
+				OnDemandQueue::<T>::set(queue);
+				return Some(Assignment::ParathreadA(entry))
+			};
+		}
+		return None
+	}
+
+	/// Pushes an assignment back to the queue on session boundaries.
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand.rs:406]: Docs are not fully accurate anymore, more accurate:

```
Push an assignment back to the queue.

Typically used on session boundaries.
```

File: runtime/parachains/src/assigner_on_demand/tests.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+			Perbill::from_percent(10),
+			Perbill::from_percent(100),
+		)
+		.unwrap()
+	}
+	assert_eq!(traffic, FixedU128::from_inner(2_718_103_312_071_174_051u128))
+}
+
+#[test]
+fn place_order_works() {
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand/tests.rs:196]: Hmm, this test seems to be mostly testing that an order does not work (but misses a few reasons).

**Review by eskimor — COMMENTED** [review]
Nice! Some more docs would be good, but very readable code, also good names!

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2804927

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2805939

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2959203

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2959321

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2969721

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2975836

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2976295

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2978179

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2997715

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3045976

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3046497

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3065209

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3070884

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+				QueuePushDirection::Front => queue.push_front(assignment),
+			};
+			Ok(())
+		});
+		res
+	}
+
+	/// Get the size of the on demand queue.
+	/// Returns:
+	/// - The size of the on demand queue, or failing to do so, the max size of the queue.
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand/mod.rs:363]: This was a bit confusing. We are not actually failing in getting the size, this is just about the corner case that it does not fit into a u32. I would suggest to ensure via the interface and tests that this can never happen and then in the `Err` case we can still use the max size, but add a debug log, saying that an invariant was not upheld. 

Then we can simply remove the comment about failing in the doc, because it should never happen and if it does it is an internal bug and the caller should not need to bother with those.

Comment (eskimor) [reply runtime/parachains/src/assigner_on_demand/mod.rs:363]: This is a private internal function so the caller is part of making sure the invariant is upheld, nevertheless let's do the debug log instead of the doc.

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+		let config = <configuration::Pallet<T>>::config();
+		match OnDemandQueue::<T>::get().len().try_into() {
+			Ok(size) => return size,
+			Err(_) => return config.on_demand_queue_max_size,
+		}
+	}
+
+	/// Getter for the order queue.
+	pub fn get_queue() -> Vec<Assignment> {
+		OnDemandQueue::<T>::get().into()
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand/mod.rs:374]: Why are we eagerly converting to a `Vec`? What is bad about the `VecDequeue`? Do the callers need a `Vec`?

**Review by eskimor — COMMENTED** [review]
Not through yet, but already have some comments. Overall really good work, comments are mostly (only?) nitpicks on how it would be even prettier/robust.

File: node/core/backing/src/tests.rs [hunk 7]
```diff
@@ -97,14 +97,11 @@ impl Default for TestState {
 		let group_rotation_info =
 			GroupRotationInfo { session_start_block: 0, group_rotation_frequency: 100, now: 1 };
 
-		let thread_collator: CollatorId = Sr25519Keyring::Two.public().into();
+		let _thread_collator: CollatorId = Sr25519Keyring::Two.public().into();
```
[the remark this answers is not in the document]
Comment (antonva) [reply node/core/backing/src/tests.rs:100]: Missed holdover from collator restrictions. Will be removed.

File: primitives/src/v5/mod.rs [hunk 8]
```diff
@@ -812,29 +812,63 @@ pub struct ParathreadClaim(pub Id, pub Option<CollatorId>);
 pub struct ParathreadEntry {
 	/// The claim.
 	pub claim: ParathreadClaim,
-	/// Number of retries.
+	/// Number of retries
 	pub retries: u32,
```
[the remark this answers is not in the document]
Comment (antonva) [reply primitives/src/v5/mod.rs:816]: Maybe shorten `attempted_tries` to just `attempts`?. After the first attempt, it gets incremented. Would want to change the `retry` naming in configuration and elsewhere to `max_attempts` to match.

File: primitives/src/v5/mod.rs [hunk 9]
```diff
@@ -812,29 +812,63 @@ pub struct ParathreadClaim(pub Id, pub Option<CollatorId>);
...
+
+/// An entry tracking a paras
+#[derive(Clone, Encode, Decode, TypeInfo, PartialEq, RuntimeDebug)]
+pub struct ParasEntry<N = BlockNumber> {
+	/// The `Assignment`
+	pub assignment: Assignment,
+	/// Number of times this has been retried.
+	pub retries: u32,
+	/// Block height when this entry was added to the claim queue.
+	pub ttl: N,
```
[the remark this answers is not in the document]
Comment (antonva) [reply primitives/src/v5/mod.rs:841]: Comment was wrong, it'll say "The block height where this entry becomes invalid". This field is still a TTL.

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+	pub fn add_parathread_assignment(
+		assignment: Assignment,
+		location: QueuePushDirection,
+	) -> Result<(), DispatchError> {
+		// Only parathreads are valid paraids for on the go parachains.
+		ensure!(<paras::Pallet<T>>::is_parathread(assignment.para_id), Error::<T>::InvalidParaId);
+
+		let config = <configuration::Pallet<T>>::config();
+
+		// Be explicit rather than implicit about the return value
```
[the remark this answers is not in the document]
Comment (antonva) [reply runtime/parachains/src/assigner_on_demand/mod.rs:348]: We agreed it shouldn't be there. I had there as a reminder it while I was reasoning about something I can't remember 😄 . Consider it gone.

File: runtime/parachains/src/scheduler_common/mod.rs [hunk 11]
```diff
@@ -51,103 +56,56 @@ pub enum FreedReason {
...
+	/// Returns the availability period specified by the implementation.
+	/// See
+	/// [`HostConfiguration::chain_availability_period`] and
+	/// [`HostConfiguration::thread_availability_period`]
+	/// for more information.
+	fn get_availability_period(core_idx: CoreIndex) -> BlockNumber;
 
+	/// How many retries a collator gets to provide a collation in this implementation.
+	/// Zero retries means that the collator has 1 chance as per the bulk (slot auction)
+	/// parachain model.
```
Comment (eskimor) [inline runtime/parachains/src/scheduler_common/mod.rs:85]: Wait if 0 means 1 try, then the naming is actually correct?

File: runtime/parachains/src/scheduler_common/mod.rs [hunk 12]
```diff
@@ -51,103 +56,56 @@ pub enum FreedReason {
...
 /// How a free core is scheduled to be assigned.
 #[derive(Clone, Encode, Decode, PartialEq, TypeInfo)]
 #[cfg_attr(feature = "std", derive(Debug))]
-pub struct CoreAssignment {
+pub struct CoreAssignment<BlockNumber> {
 	/// The core that is assigned.
 	pub core: CoreIndex,
-	/// The kind of the assignment.
-	pub kind: Assignment,
+	/// The parachain information needed to ba
```
Comment (eskimor) [inline runtime/parachains/src/scheduler_common/mod.rs:95]: needed to ba?

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+		OnDemandQueue::<T>::get().into()
+	}
+
+	/// Decreases the affinity of a `ParaId` to a specified `CoreIndex`.
+	/// Subtracts from the count of the `CoreAffinityCount` if an entry is found and the core_idx matches.
+	/// When the count reaches 0, the entry is removed.
+	/// A non-existant entry is a no-op.
+	fn decrease_affinity(para_id: ParaId, core_idx: CoreIndex) {
+		match ParaIdAffinity::<T>::get(para_id) {
+			Some(affinity) =>
```
[the remark this answers is not in the document]
Comment (antonva) [reply runtime/parachains/src/assigner_on_demand/mod.rs]: Changed both increase/decrease fns to use mutate.

File: runtime/parachains/src/assigner_on_demand/mod.rs (hunk 6 above)
Comment (antonva) [reply runtime/parachains/src/assigner_on_demand/mod.rs]: There used to be a caller that expected a `Vec`. This has been changed to a `VecDeque` as the only remaining caller for this fn is a test.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3106626

File: runtime/parachains/src/scheduler.rs [hunk 14]
```diff
@@ -59,7 +59,7 @@ pub use pallet::*;
 #[cfg(test)]
 mod tests;
 
-const LOG_TARGET: &str = "runtime::parachain_scheduler";
+const LOG_TARGET: &str = "runtime::parachains::scheduler";
```
Comment (eskimor) [inline runtime/parachains/src/scheduler.rs:62]: Why was the log target changed?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3130772

File: runtime/parachains/src/configuration.rs [hunk 15]
```diff
@@ -297,9 +301,7 @@ pub enum InconsistentError<BlockNumber> {
 	/// `group_rotation_frequency` is set to zero.
 	ZeroGroupRotationFrequency,
 	/// `chain_availability_period` is set to zero.
```
Comment (tdimitrov) [inline runtime/parachains/src/configuration.rs:303]:
```suggestion
	/// `paras_availability_period` is set to zero.
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3236468

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3236688

File: runtime/polkadot/src/weights/runtime_parachains_configuration.rs [hunk 16]
```diff
@@ -150,4 +150,6 @@ impl<T: frame_system::Config> runtime_parachains::configuration::WeightInfo for
 			.saturating_add(T::DbWeight::get().reads(4))
 			.saturating_add(T::DbWeight::get().writes(1))
 	}
+
+	fn set_config_with_perbill() -> Weight { todo!() }
```
Comment (tdimitrov) [inline runtime/polkadot/src/weights/runtime_parachains_configuration.rs:154]: The implementation here is missing.

File: primitives/src/v5/mod.rs [hunk 17]
```diff
@@ -966,10 +1000,9 @@ impl<H, N> OccupiedCore<H, N> {
 #[derive(Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
 #[cfg_attr(feature = "std", derive(PartialEq))]
 pub struct ScheduledCore {
+	// TODO: Is the same as Assignment
```
Comment (tdimitrov) [inline primitives/src/v5/mod.rs:1003]: A forgotten TODO?

File: runtime/rococo/src/weights/runtime_parachains_configuration.rs [hunk 18]
```diff
@@ -148,4 +148,6 @@ impl<T: frame_system::Config> runtime_parachains::configuration::WeightInfo for
 			.saturating_add(T::DbWeight::get().reads(4))
 			.saturating_add(T::DbWeight::get().writes(1))
 	}
+
+	fn set_config_with_perbill() -> Weight { todo!() }
```
Comment (tdimitrov) [inline runtime/rococo/src/weights/runtime_parachains_configuration.rs:152]: Missing implementation.

(I know you are working on this atm. I'm marking it for convenience.)

File: runtime/rococo/src/lib.rs [hunk 19]
```diff
@@ -879,6 +881,7 @@ pub enum ProxyType {
 	CancelProxy,
 	Auction,
 	Society,
+	OnDemandOrdering,
```
Comment (tdimitrov) [inline runtime/rococo/src/lib.rs:884]: Out of curiosity - why this proxy type exists only on Rococo?

File: runtime/westend/src/weights/runtime_parachains_configuration.rs [hunk 20]
```diff
@@ -140,4 +140,6 @@ impl<T: frame_system::Config> runtime_parachains::configuration::WeightInfo for
 			.saturating_add(T::DbWeight::get().reads(3))
 			.saturating_add(T::DbWeight::get().writes(1))
 	}
+
+	fn set_config_with_perbill() -> Weight { todo!() }
```
Comment (tdimitrov) [inline runtime/westend/src/weights/runtime_parachains_configuration.rs:144]: Ditto (missing implementation)

File: runtime/parachains/src/assigner.rs [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+//! The Polkadot multiplexing assignment provider.
+//! Provides blockspace assignments for both bulk and on demand parachains.
+use frame_system::pallet_prelude::BlockNumberFor;
+use primitives::{v5::Assignment, CoreIndex, Id as ParaId};
+
+use crate::{configuration, paras, scheduler_common::AssignmentProvider};
+
+pub use pallet::*;
+
+//#[cfg(test)]
```
Comment (tdimitrov) [inline runtime/parachains/src/assigner.rs:26]: Should this be removed?

File: runtime/parachains/src/scheduler_common/mod.rs [hunk 22]
```diff
@@ -52,103 +56,56 @@ pub enum FreedReason {
...
+	/// that need to be able to keep track of assignments over session boundaries,
+	/// such as the on demand assignment provider.
 	fn push_assignment_for_core(core_idx: CoreIndex, assignment: Assignment);
 
-	fn get_availability_period(core_idx: CoreIndex) -> BlockNumberFor<T>;
+	/// Returns the availability period specified by the implementation.
+	/// See
+	/// [`HostConfiguration::chain_availability_period`] and
+	/// [`HostConfiguration::thread_availability_period`]
+	/// for more information.
```
Comment (tdimitrov) [inline runtime/parachains/src/scheduler_common/mod.rs:80]:
```suggestion
	/// [`HostConfiguration::paras_availability_period`]
	/// for more information.
```

File: runtime/parachains/src/scheduler.rs [hunk 23]
```diff
@@ -113,14 +113,21 @@ pub mod pallet {
 	pub(crate) type SessionStartBlock<T: Config> = StorageValue<_, BlockNumberFor<T>, ValueQuery>;
 
 	/// One entry for each availability core. The `VecDeque` represents the assignments to be scheduled on that core.
+	/// `None` is used to signal to not schedule the next para of the core as there is one currently being scheduled.
```
Comment (tdimitrov) [inline runtime/parachains/src/scheduler.rs:116]: I don't understand this line. How is made the relation between `Para` and `None`?

File: runtime/parachains/src/scheduler.rs [hunk 24]
```diff
@@ -113,14 +113,21 @@ pub mod pallet {
...
+	/// Not using `None` here would overwrite the `CoreState` in the runtime API.
+
+	// TODO: this is behaviour is likely going to change when adapting for AB
```
Comment (tdimitrov) [inline runtime/parachains/src/scheduler.rs:119]: Should this be removed?

File: runtime/parachains/src/scheduler.rs (hunk 14 above)
Comment (antonva) [reply runtime/parachains/src/scheduler.rs:62]: The log target isn't set in the scheduler prior to the scheduling changes. This brings it in line with what I am using in `assigner_on_demand` as well as the `slashing` module.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3263357

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 25]
```diff
@@ -0,0 +1,534 @@
...
+use sp_std::{collections::vec_deque::VecDeque, prelude::*};
+
+const LOG_TARGET: &str = "runtime::parachains::on-demand-assigner";
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod tests;
+
+// TODO enable benchmarks maybe?
```
Comment (tdimitrov) [inline runtime/parachains/src/assigner_on_demand/mod.rs:54]: TODO...

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 26]
```diff
@@ -0,0 +1,534 @@
...
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+
+pub trait WeightInfo {
+	fn place_order() -> Weight;
+}
+
+/// A weight info that is only suitable for testing.
+pub struct TestWeightInfo;
```
Comment (tdimitrov) [inline runtime/parachains/src/assigner_on_demand/mod.rs:63]:
```suggestion
#[cfg(test)]
pub struct TestWeightInfo;
```

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 27]
```diff
@@ -0,0 +1,534 @@
...
+
+impl WeightInfo for TestWeightInfo {
```
Comment (tdimitrov) [inline runtime/parachains/src/assigner_on_demand/mod.rs:65]:
```suggestion
#[cfg(test)]
impl WeightInfo for TestWeightInfo {
```

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 28]
```diff
@@ -0,0 +1,534 @@
...
+		}
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	BalanceOf<T>: FixedPointOperand,
+{
+	/// The spot price multiplier. This is based on the transaction fee calculations defined in:
+	/// https://research.web3.foundation/en/latest/polkadot/overview/2-token-economics.html#setting-transaction-fees
```
Comment (tdimitrov) [inline runtime/parachains/src/assigner_on_demand/mod.rs:301]:
```suggestion
	/// https://research.web3.foundation/Polkadot/overview/token-economics#setting-transaction-fees
```

The link has changed for some reason

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3272949

File: runtime/parachains/src/scheduler/tests.rs [hunk 29]
```diff
@@ -116,6 +113,81 @@ fn default_config() -> HostConfiguration<BlockNumber> {
 	}
 }
 
+#[cfg(test)]
+pub(crate) fn claimqueue_contains_only_none() -> bool {
+	let mut cq = Scheduler::claimqueue();
+	//let mut cq = ClaimQueue::<T>::get();
```
Comment (tdimitrov) [inline runtime/parachains/src/scheduler/tests.rs:119]:
```suggestion
```

File: runtime/parachains/src/scheduler/tests.rs [hunk 30]
```diff
@@ -116,6 +113,81 @@ fn default_config() -> HostConfiguration<BlockNumber> {
 	}
 }
 
+#[cfg(test)]
```
Comment (tdimitrov) [inline runtime/parachains/src/scheduler/tests.rs:116]: Nit: the whole module is guarded by `#[cfg(test)]` so these here can be removed.

File: runtime/parachains/src/assigner_on_demand/tests.rs (hunk 4 above)
Comment (antonva) [reply runtime/parachains/src/assigner_on_demand/tests.rs:207]: Test now does what it says on the tin, albeit with checks along the way.

File: runtime/parachains/src/scheduler.rs [hunk 31]
```diff
@@ -487,77 +517,106 @@ impl<T: Config> Pallet<T> {
...
+					if Self::is_core_occupied(core_idx) { 1 } else { 0 };
 				for _ in n_lookahead_used..n_lookahead {
 					let concluded_para = concluded_paras.remove(&core_idx);
-					if let Some(ass) =
+					if let Some(assignment) =
 						T::AssignmentProvider::pop_assignment_for_core(core_idx, concluded_para)
 					{
-						let ca = ass.to_core_assignment(core_idx, group_idx);
-						Self::add_to_claimqueue(ca);
+						let ttl = <configuration::Pallet<T>>::config().on_demand_ttl;
```
Comment (eskimor) [inline runtime/parachains/src/scheduler.rs:542]: Oh oh. This looks off. We are querying the ttl value for on_demand, but the core could be occupied by anything. In the scheduler we should not know anything about on-demand.

Comment (eskimor) [reply runtime/parachains/src/scheduler.rs:542]: If there is no other way, this needs to be a function of the `AssignmentProvider` trait. If we go that route, maybe we should combine all those functions into a `core_config()` function or something.

File: runtime/rococo/src/weights/runtime_parachains_assigner_on_demand.rs [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+/// Weight functions for `runtime_parachains::assigner_on_demand`.
+pub struct WeightInfo<T>(PhantomData<T>);
+impl<T: frame_system::Config> runtime_parachains::assigner_on_demand::WeightInfo for WeightInfo<T> {
+	/// Storage: `OnDemandAssignmentProvider::SpotTraffic` (r:1 w:0)
+	/// Proof: `OnDemandAssignmentProvider::SpotTraffic` (`max_values`: Some(1), `max_size`: None, mode: `Measured`)
+	/// Storage: `Paras::ParaLifecycles` (r:1 w:0)
+	/// Proof: `Paras::ParaLifecycles` (`max_values`: None, `max_size`: None, mode: `Measured`)
+	/// Storage: `OnDemandAssignmentProvider::OnDemandQueue` (r:1 w:1)
+	/// Proof: `OnDemandAssignmentProvider::OnDemandQueue` (`max_values`: Some(1), `max_size`: None, mode: `Measured`)
+	fn place_order() -> Weight {
```
Comment (eskimor) [inline runtime/rococo/src/weights/runtime_parachains_assigner_on_demand.rs:56]: :+1:

**Comment by eskimor** [comment]
bot bench $ runtime rococo runtime_parachains::assigner_on_demand

**Comment by eskimor** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_parachains::assigner_on_demand

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3331601

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3361467

File: node/network/collator-protocol/src/validator_side/mod.rs [hunk 33]
```diff
@@ -959,6 +942,24 @@ async fn process_incoming_peer_message<Context>(
 				Some(p) => p,
 			};
 
+			if !state.view.contains(&relay_parent) {
```
[the remark this answers is not in the document]
Comment (antonva) [reply node/network/collator-protocol/src/validator_side/mod.rs:945]: This looks to me like a vestigial change from when we had a `can_collate` fn that depended on `peer_data`. Now that we've removed the collator restrictions it should probably go back to the way it was.

**Comment by antonva** [comment]
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_parachains::configuration

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3366192

**Comment by antonva** [comment]
bot bench polkadot-pallet --runtime=kusama --pallet=runtime_parachains::configuration
bot bench polkadot-pallet --runtime=polkadot --pallet=runtime_parachains::configuration
bot bench polkadot-pallet --runtime=westend --pallet=runtime_parachains::configuration
bot bench polkadot-pallet --runtime=rococo --pallet=runtime_parachains::assigner_on_demand

File: runtime/parachains/src/assigner_on_demand/mod.rs [hunk 34]
```diff
@@ -55,14 +56,14 @@ const LOG_TARGET: &str = "runtime::parachains::on-demand-assigner";
 pub use pallet::*;
 
 pub trait WeightInfo {
-	fn place_order() -> Weight;
+	fn place_order(s: u32) -> Weight;
```
Comment (eskimor) [inline runtime/parachains/src/assigner_on_demand/mod.rs:59]: I think this needs to be adjusted as well.

**Review by eskimor — COMMENTED** [review]
Great work Anton!

**Comment by eskimor** [comment]
Let's merge this into the feature branch for good measure. The one remaining fix can be done on the feature branch directly.
